### PR TITLE
Ban prototype pollution and document it

### LINKED ISSUE #216: Prevent prototype injection
It appears fast-json-patch may be vulnerable to prototype injection depending on the way it is used. See e.g. the following code:

````js
let fp = require("fast-json-patch");

function SomeClass() {
        this.foo = "bar";
}

let doc = new SomeClass();
let otherDoc = new SomeClass();

const patch = [
        {op: "replace", path: "/__proto__/x", value: "Hacked!"}
];

fp.applyPatch(doc, patch);
console.log(doc.x, otherDoc.x);
````

This will output 'hacked' twice, while there is no property `x` visible on object `doc` nor `otherDoc` when it is printed.

This becomes an issue in scenarios where patches cannot be trusted (and are not checked for this). Perhaps it is a good idea to disable patching of `__proto__` (and maybe `prototype` as well? I haven't tested that yet) unless enabled by a flag (although I do not really see a use case where you would use JSON patches to fix up a prototype...).

### PR DESCRIPTION
**Pull request by alshakero**
fixes #216

### CODE REVIEW

File: dist/fast-json-patch.js [hunk 1]
```diff
@@ -504,7 +511,7 @@ function applyPatch(document, patch, validateOperation, mutateDocument) {
     }
     var results = new Array(patch.length);
     for (var i = 0, length_1 = patch.length; i < length_1; i++) {
-        results[i] = applyOperation(document, patch[i], validateOperation);
+        results[i] = applyOperation(document, patch[i], validateOperation, true, banPrototypeModifications);
```
Comment (tomalec) [inline dist/fast-json-patch.js]: Cosmetics,
You can consider explaining why you hard-code one parameter and forward the others, as it caused me to think for a second.
Like, "`mutateDocument` was already covered for the entire sequence, we will apply operations on cloned document if applicable"

File: test/spec/coreSpec.js [hunk 2]
```diff
@@ -1910,5 +1916,56 @@ describe('undefined - JS to JSON projection / JSON to JS extension', function()
         bar: null
       });
     });
+
+    it(`should allow __proto__ modifications when the flag is set`, function() {
```
Comment (tomalec) [inline test/spec/coreSpec.js]: To me more descriptive
```suggestion
    it(`should allow __proto__ modifications when the mutateDocument flag is set`, function() {
```

File: test/spec/coreSpec.js [hunk 3]
```diff
@@ -1910,5 +1916,56 @@ describe('undefined - JS to JSON projection / JSON to JS extension', function()
...
+      const patch = [
+        { op: 'replace', path: `/__proto__/x`, value: 'polluted' }
+      ];
+
+      jsonpatch.applyPatch(doc, patch, false, true, false);
+
+      expect(otherDoc.x).toEqual('polluted');
+    });
+
+    it(`should not allow __proto__ modifications without setting the flag and should throw an error`, function() {
```
Comment (tomalec) [inline test/spec/coreSpec.js]:
```suggestion
    it(`should not allow __proto__ modifications without setting the mutateDocument flag and should throw an error`, function() {
```

File: test/spec/coreSpec.js [hunk 4]
```diff
@@ -1910,5 +1916,56 @@ describe('undefined - JS to JSON projection / JSON to JS extension', function()
...
+
+      const patch = [
+        { op: 'replace', path: `/__proto__/x`, value: 'polluted' }
+      ];
+
+      try {
+        jsonpatch.applyPatch(doc, patch);
+      } catch (e) {
+        expect(e.message).toEqual(expectedErrorMessage);
+      }
```
Comment (tomalec) [inline test/spec/coreSpec.js]: Have you considered `expect(()=>{jsonpatch.applyPatch(doc, patch)}).to.throw(TypeError, expectedErrorMessage);` to make test easier to read, and check for the error type as well?
https://www.chaijs.com/api/bdd/#method_throw

I'm afraid the code above would pass the test if `jsonpatch.applyPatch(doc, patch);` does not throw at all, as then the `expect` function is not called either.

**Comment by warpech** [comment]
Please consider that PR#221 also proposes to change the signature of the method `applyOperation`

**Comment by alshakero** [comment]
Addressed all and tests are passing. merging.

**Pull request merged**